### PR TITLE
dma: Refactor channel operations interface

### PR DESCRIPTION
**Pull request by akloniex**
Extract common part of *_dma_chan_data into dma_chan_data structure with private pointer for class-specific data access, and expose it to the consumer for later use throughout lifecycle.
This is done in order to simplify API, as well as a preparation for future implementation of callbacks system for which it would be beneficial to have separate 'event caller' instances for each of the DMA channels for consumer to register to, as opposed to current, single DMA 'caller' exposed for registration.

Depends on fix: https://github.com/thesofproject/sof/pull/1751

### CODE REVIEW

File: src/platform/apollolake/include/platform/lib/dma.h [hunk 1]
```diff
@@ -47,7 +47,7 @@
 #define DMA_HANDSHAKE_SSP5_TX	12
 #define DMA_HANDSHAKE_SSP5_RX	13
 
-#define dma_chan_irq(dma, chan) (dma_irq(dma) + channel)
+#define dma_chan_irq(dma, chan) (dma_irq(dma) + chan)
 
 int dmac_init(void);
```
Comment (dbaluta) [inline src/platform/apollolake/include/platform/lib/dma.h]: @akloniex this is an obvious bug. You can even send it as a separate pull request to merge it quicker.

Comment (paulstelian97) [reply src/platform/apollolake/include/platform/lib/dma.h]: Must have worked from the fact that the macro was called always with the name "channel"... Funny how some of these issues just slip like that.

Comment (tlauda) [reply src/platform/apollolake/include/platform/lib/dma.h]: @dbaluta @paulstelian97 It's already in separate PR #1751.

Comment (dbaluta) [reply src/platform/apollolake/include/platform/lib/dma.h]: @tlauda right. Merged!

Comment (jajanusz) [reply src/platform/apollolake/include/platform/lib/dma.h]: That's why macros suck. I don't understand love for them in c projects. Static inline functions are way better.

Comment (dbaluta) [reply src/platform/apollolake/include/platform/lib/dma.h]: @jajanusz move fast breaks things! For simple constructs like the one above macros are just OK. That's the beauty of C language. You can never stop having fun with it! :)

File: src/drivers/dw/dma.c [hunk 2]
```diff
@@ -68,20 +65,11 @@ struct dw_dma_chan_data {
...
-	/* callback type */
-	int cb_type;
-
-	struct dma *dma;
 };
 
 /* private data for DW DMA engine */
 struct dma_pdata {
-	struct dw_dma_chan_data chan[DW_MAX_CHAN];
+	struct dma_chan_data chan[DW_MAX_CHAN];
```
Comment (tlauda) [inline src/drivers/dw/dma.c]: Now it is a common struct, so don't need to be a part of dma private data.
It can be a dma struct member.

Comment (paulstelian97) [reply src/drivers/dw/dma.c]: DW_MAX_CHAN is kinda specific, and you can't use constants in the platform data to decide the size. So how should he do this?

Comment (tlauda) [reply src/drivers/dw/dma.c]: Or we could merge it with the one in platform data and allocate it dynamically based on number of channels assigned to the particular DMA.

Comment (paulstelian97) [reply src/drivers/dw/dma.c]: @tlauda Yeah, dynamic allocation is also an option. Let's just hope it doesn't lead to allocating more than 1024 bytes in one allocation (I have encountered issues with that while working on the almost-ready EDMA driver on i.MX).

Comment (akloniex) [reply src/drivers/dw/dma.c]: @tlauda I've extracted channels to dma struct and initialize them during dma_probe.

**Comment by paulstelian97** [comment]
@tlauda I'm actually interested in this one getting merged since I'm in the process of writing and testing the EDMA driver (for i.MX). The changes are significant enough to make me rewrite quite a big portion of it and will also help me test it. It's also got updated with your latest comments so can you please re-review? Thanks.

**Comment by dbaluta** [comment]
@akloniex good work! This is a nice PR.

File: src/drivers/dw/dma.c [hunk 3]
```diff
@@ -1290,6 +1279,18 @@ static int dw_dma_probe(struct dma *dma)
 		return -ENOMEM;
 	}
 	dma_set_drvdata(dma, dw_pdata);
+#endif
+
+	/* allocate dma channels */
+	dma->chan = rzalloc(RZONE_SYS_RUNTIME | RZONE_FLAG_UNCACHED,
+			    SOF_MEM_CAPS_RAM,
+			    sizeof(struct dma_chan_data) * DW_MAX_CHAN);
```
Comment (tlauda) [inline src/drivers/dw/dma.c]: We don't need to allocate for `DW_MAX_CHAN` anymore. Now we can read actual number of channels from `dma->plat_data.channels`.

File: src/drivers/intel/cavs/hda-dma.c [hunk 4]
```diff
@@ -816,61 +800,84 @@ static int hda_dma_set_cb(struct dma *dma, unsigned int channel, int type,
...
 		return -EEXIST; /* already created */
 
-	/* allocate private data */
-	hda_pdata = rzalloc(RZONE_SYS_RUNTIME | RZONE_FLAG_UNCACHED,
-			    SOF_MEM_CAPS_RAM, sizeof(*hda_pdata));
-	if (!hda_pdata) {
-		trace_hddma_error("hda-dmac: %d alloc failed",
+	dma->chan = rzalloc(RZONE_SYS_RUNTIME | RZONE_FLAG_UNCACHED,
+			    SOF_MEM_CAPS_RAM,
+			    sizeof(struct dma_chan_data) * HDA_DMA_MAX_CHANS);
```
Comment (tlauda) [inline src/drivers/intel/cavs/hda-dma.c]: `dma->plat_data.channels`

File: src/drivers/dw/dma.c (hunk 3 above)
Comment (akloniex) [reply src/drivers/dw/dma.c]: @tlauda done

File: src/drivers/intel/cavs/hda-dma.c (hunk 4 above)
Comment (akloniex) [reply src/drivers/intel/cavs/hda-dma.c]: @tlauda done

**Comment by tlauda** [comment]
@akloniex It's OK now, but CI still massively failing.

File: src/drivers/intel/cavs/hda-dma.c [hunk 5]
```diff
@@ -212,7 +201,8 @@ static void hda_dma_get_dbg_vals(struct hda_chan_data *chan,
 
 #define hda_dma_ptr_trace(chan, postfix, src) \
 	do { \
-		struct hda_dbg_data *dbg_data = &(chan)->dbg_data; \
+		struct hda_chan_data *hda_chan = dma_chan_get_data(chan); \
+		struct hda_dbg_data *dbg_data = &(hda_chan)->dbg_data; \
```
Comment (paulstelian97) [inline src/drivers/intel/cavs/hda-dma.c]: Parentheses not needed around hda_chan (it's not a macro parameter). Doesn't explain CI failures though...

File: src/drivers/intel/cavs/hda-dma.c [hunk 6]
```diff
@@ -802,146 +723,156 @@ static int hda_dma_pm_context_store(struct dma *dma)
...
 
-	for (i = 0; i < HDA_DMA_MAX_CHANS; i++, chan++) {
+	for (i = 0; i < dma->plat_data.channels; i++, chan++) {
 		chan->dma = dma;
 		chan->index = i;
 		chan->status = COMP_STATE_INIT;
+
+		hda_chan = rzalloc(RZONE_SYS_RUNTIME | RZONE_FLAG_UNCACHED,
+				   SOF_MEM_CAPS_RAM,
+				   sizeof(struct hda_chan_data));
```
Comment (paulstelian97) [inline src/drivers/intel/cavs/hda-dma.c]: Could someone actually do some tests? Namely add a print if this allocation fails (might be this the cause of CI failures?)

Comment (paulstelian97) [reply src/drivers/intel/cavs/hda-dma.c]: @xiulipan Would any trace_error calls be visible in the SOFCI tests on HDA platforms?

File: src/drivers/dw/dma.c [hunk 7]
```diff
@@ -54,11 +54,8 @@ struct dw_dma_ptr_data {
 
 /* data for each DW DMA channel */
 struct dw_dma_chan_data {
-	uint32_t status;
```
Comment (lgirdwood) [inline src/drivers/dw/dma.c]: Please add the details of what is being changed in the commit message.

File: src/drivers/intel/cavs/hda-dma.c (hunk 6 above)
Comment (xiulipan) [reply src/drivers/intel/cavs/hda-dma.c]: @paulstelian97 CI will not find any trace_error if there is no error happen.
But we can run some test.

BTW: are you asking the error logs for this PR or for code base?

Comment (paulstelian97) [reply src/drivers/intel/cavs/hda-dma.c]: @xiulipan I would have suggested to add a trace_error temporarily that triggers if rzalloc failed, as I suspect it may be at the root of the issue with all HDA failures.

File: src/audio/dai.c [hunk 8]
```diff
@@ -55,7 +55,7 @@
```
Comment (dbaluta) [inline src/audio/dai.c]: @akloniex Please explain in the commit message why this commit is needed. I understand that this is very clear for you now, but people reading the commit message history one year from now will be very confused.

**Comment by tlauda** [comment]
SOFCI TEST

**Comment by tlauda** [comment]
SOFCI TEST

**Comment by paulstelian97** [comment]
@tlauda @akloniex Any suggestions for platforms with 32 channels where the array of descriptors would become larger than the 1024-byte block size of the allocator? Should I add more, larger blocks on my platform? (I have enough memory for that) Or should I somehow split that allocation?

**Comment by tlauda** [comment]
@paulstelian97 This is why we have memory map per platform. Everyone can have custom settings appropriate for their usage scenarios. If you need to add larger blocks, then add it.
